### ISSUE TITLE
Formalize communication interface between netmaster and netplugin

### ISSUE THREAD

**Issue by mapuri** [contiv/netplugin#51]
Right now the netplugin daemon get's triggered in two ways:
- **etcd events** generated as a result of the configuration created by netmaster. These events in turn trigger the netplugin to invoke underlying driver implementation that programs the data path in host.
- **events from docker daemon** generated on container start. These events in turn trigger the netplugin to perform container net-namespace manipulations (like ovs ports to container namespace and assigning IP addresses)

This issue track removing the dependence on **etcd events** and instead formalize it as a RPC like interface. This decouples the netplugin daemon from depending on differing implementations of the events in different data-stores in the deployment environment.

Eventually, we might need to define an RPC interface to replace docker events for reasons like being integration with native docker hooks etc. But that can be handled as separate PR.

**Comment by mapuri** [comment]
**TL;DR**

- Defining a communication interface tailored for netmaster and netplugin requires certain patterns of a distributed systems to be implemented like discovery/clustering, unicast and multicast messaging support etc.
- the current state based interface does offer multiple characteristic of desired communication patterns using even state monitoring, however, we can enhance the `core.State` and `core.Statedriver` to support a method like `watch` to formalize this dependency.
- Given some of the foreseeable deployments of netplugin shall be in an environment that use distributed state (K8s, docker-swarm, mesos etc), it might be ok to keep a state based interface.
- We can keep the interface as is till we find a more compelling usecase that wouldn't fit the State based communication environment.

++++++++
Following are some of the thoughts and approaches about changing/enhancing the interface between netmaster and netplugin that I am figuring out as part of investigating the changes for the original issue.

The communication requirements in this issue can be summarized as following:
- netmaster shall be able to unicast messages meant for a specific netplugin.
 - this communication can be two way i.e. netmaster can initiate a request to netplugin and vice versa.
 - *Are there cases where netplugin instances need to talk to each other?*
- netmaster shall be able to multicast messages meant for all/more than one netplugin instances.
 - this communication shall be originated by the master.
 - *Are there cases where a netplugin instance needs to originate multicast messages?*
- to reach a netplugin instance the netmaster needs to have a registration/discovery mechanism to allow the netplugin instances to express interest in receiving notifications.
- to reach all/more than one netplugin instances netmaster needs to have ability to deliver a copy to each instance.
- communication needs to be reliable (retries, acks, replay in case of a netplugin instance starting up first time or after a crash etc)

The above are some common patterns that are provided/supported by multiple distributed systems and inline with that we will also need concepts like discovery mechanisms similar to systems like etcd, K8s etc and communication mechanisms provided by libraries like zmq.

However, it deems suitable to compare the current communication support with the new requirements wrt the gains and challenges of each. The comparison assumes the following, which is mostly true in current implementation:
- a netplugin instance implements host/minion local functions like handling container up/down events; handling configuration events from netmaster and programming the host.
 - a netplugin instance runs on each minion/host
 - a minion/host is a part of a resource cluster that is managed by a (usually distributed) management function and runs some form of agent of that function, like it runs where kubelet run in a K8s deployment; it runs where docker daemon runs in a docker swarm deployment.
- netmaster implements the logically centralized functions like consuming logical (ops) intent and translating it to concrete (network implementation specific) intent; providing central resource allocation; handling endpoint to host binding etc
 - netmaster runs from or as part of the (logically) central entity of the distributed management function.

The two communication mechanisms can be compared as follows:

- **State based**, as it is currently. This approach assumes that all/some system state is stored in a *distributed* state store (like etcd, consul etc) that provides a mechanism to notify state changes.

  **What it provides/Gains:**
  - provides the multicast characteristics of communication pattern: 
    - unicast can be implemented as a special case by organizing/structuring the state such that even if the state change trigger reaches each netplugin instance only the intended instance shall act.
  - provides reliability of communication: it's easy to restore state on startup or after crash from data store; state change notifications can be used to trigger read of latest state; 
  
  **What it lacks/Challenges:**
  - depends on certain features from the distributed state store like state-change events; reliability/consistency of state reads.
    - state-change events might be easily addressed by implementing wrapper to mimic this functionality on top of distributed store, like a state polling loop like adding a `watch` method to the `core.State` and `core.Statedriver` interfaces.
  - might be a bit inefficient since it requires using state change events (multicast) even for unicast events.
    - it might not be a big issue as we are talking an instance per host.

- **RPC based**, as being proposed.

  **What it provides/Gains:**
  - no dependence on state ensures that the interface between netmaster and neplugin can stay independent of varied state driver and State implementations and can be optimized independently. 
  
  **What it lacks/Challenges:**
  - requires implementing yet another layer of registration; keep-alives; multicast/unicast communication between netmaster and (possibly several instance of) netplugin.
